$DragonFly: src/usr.sbin/ntpd/patches/Attic/client.c.patch,v 1.4 2005/04/14 11:22:16 joerg Exp $

--- client.c	25 Feb 2005 16:43:19 -0000	1.6
+++ client.c	12 Apr 2005 20:13:45 -0000
@@ -155,7 +155,7 @@
 
 	p->query->msg.xmttime.int_partl = arc4random();
 	p->query->msg.xmttime.fractionl = arc4random();
-	p->query->xmttime = gettime();
+	p->query->xmttime = gettime_corrected();
 
 	if (ntp_sendmsg(p->query->fd, NULL, &p->query->msg,
 	    NTP_MSGSIZE_NOAUTH, 0) == -1) {
@@ -190,7 +190,7 @@
 			fatal("recvfrom");
 	}
 
-	T4 = gettime();
+	T4 = gettime_corrected();
 
 	ntp_getmsg(buf, size, &msg);
 
@@ -228,6 +228,14 @@
 
 	p->reply[p->shift].offset = ((T2 - T1) + (T3 - T4)) / 2;
 	p->reply[p->shift].delay = (T4 - T1) - (T3 - T2);
+	if (p->reply[p->shift].delay < 0) {
+		interval = error_interval();
+		set_next(p, interval);
+		log_info("reply from %s: negative delay %f",
+		    log_sockaddr((struct sockaddr *)&p->addr->ss),
+		    p->reply[p->shift].delay);
+		return (0);
+	}
 	p->reply[p->shift].error = (T2 - T1) - (T3 - T4);
 	p->reply[p->shift].rcvd = time(NULL);
 	p->reply[p->shift].good = 1;
